File: insertion-sort.py
# Introduction to Algorithms - 3rd Edition
# page 16

# actual   = [11, 10, -1, 5, 15, 3, 5, 13]
# expected = [-1, 3, 5, 5, 10, 11, 13, 15]
def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

    return arr

# actual   = [11, 10, -1, 5, 15, 3, 5, 13]
# expected = [15, 13, 11, 10, 5, 5, 3, -1]
def insertion_sort_non_increasing(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] < key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

    return arr

if __name__ == '__main__':
    arr = [11, 10, -1, 5, 15, 3, 5, 13]
    print('increasing')
    print(insertion_sort(arr))

    arr2 = [11, 10, -1, 5, 15, 3, 5, 13]
    print('decreasing')
    print(insertion_sort_non_increasing(arr2))

File: golang/sorting/insertion_sort.go
package main

import (
	"fmt"
)

func main() {
	var arr []int8 = []int8{1, 5, -1, 0, 7, 15, 7, 10}

	insertionSort(arr)

	fmt.Println(arr)
}

// O(n^2)
func insertionSort(arr []int8) {
	var slice_size int = len(arr) // O(1)
	for i := 1; i < slice_size; i++ {
		var key int8 = arr[i]
		var j int = i - 1
		for j >= 0 && arr[j] > key {
			arr[j+1] = arr[j]
			j--
		}
		arr[j+1] = key
	}
}
